README.rst
Connexions Logging Application / Utility
========================================

This is a simple interface for logging client-side metrics and messages.

The application/utility is used to proxy client-side metrics and messages
to a running ``statsd`` service and logging service, respectively.
If a statsd service has not been configured,
the metrics will default to logging.

There are two ways to use this codebase:

1) As a standalone application which accepts and incoming request.
2) Within an existing application (recommended).

Security is handled via an HTTP header ``X-cnx-logging-key``,
which a shared key to be configured on both sides.
This is useful in the case where two or more applications share the service,
whicha also enables the applications to handle authnz in their own way.


Getting started
---------------

This is built as a Python WSGI application as well as an importable utility.

Installing
~~~~~~~~~~

To install the application itself::

    pip install .

Running the standalone application
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reinstall the application with the following::

    pip install .[standalone]

To run the application, use the ``paste`` script with the ``serve`` command.
(The paste script and serve command come from ``PasteScript`` and
``PasteDeploy``, respectively.)

This example uses the ``development.ini``, which has been supplied with the
package.  If you changed any of the database setup values, you'll also need to
change them in the configuration file.::

    paster serve development.ini

You can then surf to the address printed out by the above command.

Running tests
-------------

.. image:: https://travis-ci.org/Connexions/cnx-logging.png?branch=master
   :target: https://travis-ci.org/Connexions/cnx-logging

The tests use the standard library ``unittest`` package
and can therefore be run with minimal effort.
::

    $ python -m unittest discover

Or::

    $ python setup.py test

API
---

:/metric: Accepts a JSON formatted POST containing
	  the metric ``type``, ``label`` and ``value``.

	  For example::

	      {"type": "incr", "label": "i.haz.clikd.cheezburgr"}

:/log: Accepts a JSON formatted POST contains a ``message``.

       For example::

	   {"message": "Smoo clikd on a cheezburgr"}

License
-------

This software is subject to the provisions of the GNU Affero General
Public License Version 3.0 (AGPL). See license.txt for details.
Copyright (c) 2013 Rice University


CHANGES.txt
0.x
---

- Added path routing for metric or log acceptance instead of doing
  the routing after data parsing. [pumazi]

0.0
---

- Initialized project from existing specification found in
  rhaptos2.repo [pumazi]


tests.py
# -*- coding: utf-8 -*-
# ###
# Copyright (c) 2013, Rice University
# This software is subject to the provisions of the GNU Affero General
# Public License version 3 (AGPLv3).
# See LICENCE.txt for details.
# ###
import unittest
import logging
import json
from wsgiref.util import setup_testing_defaults

import statsd


class MockStatist(statsd.StatsClient):

    def __init__(self, test_case):
        self._prefix = None
        self.test_case = test_case

    def _send(self, data):
        if not hasattr(self.test_case, 'stats'):
            self.test_case.stats = []
        self.test_case.stats.append(data)


class LogCapturingHandler(logging.Handler):

    def __init__(self, test_case, level=logging.NOTSET):
        self.test_case = test_case
        super(LogCapturingHandler, self).__init__(level)

    def emit(self, record):
        if not hasattr(self.test_case, 'logged'):
            self.test_case.logged = []
        self.test_case.logged.append(record)


class StatistCase(unittest.TestCase):

    def test_creation_of_statsd_client(self):
        settings = {
            'statsd.host': 'example.com',
            'statsd.port': '8125',
            'statsd.prefix': '',
            }
        from cnxlogging import make_statist
        statist = make_statist(settings)
        import statsd
        self.assertTrue(isinstance(statist, statsd.StatsClient))

    def test_creation_degrades_to_logging(self):
        settings = {}
        from cnxlogging import make_statist
        statist = make_statist(settings)
        from cnxlogging import _StatsLoggingClient
        self.assertTrue(isinstance(statist, _StatsLoggingClient))


class ApplicationTestCase(unittest.TestCase):

    def setUp(self):
        self.statist = MockStatist(self)
        logger = logging.getLogger(self.__class__.__name__)
        logger.setLevel(logging.INFO)
        self.log_handler = LogCapturingHandler(self)
        logger.addHandler(self.log_handler)
        from cnxlogging import Application
        self.app = Application(self.statist, logger)

    def start_response(self, *args, **kwargs):
        self.resp_args = args
        self.resp_kwargs = kwargs

    def test_metric_acceptance(self):
        metric_data = {
            'type': 'incr',
            'label': 'i.haz.clikd.cheezburgr',
            ##'value': None,  # Tests that missing value or `None' becomes 1.
            }
        environ = {
            'PATH_INFO': '/metric',
            'REQUEST_METHOD': 'POST',
            }
        setup_testing_defaults(environ)
        # Assign the posted message.
        environ['wsgi.input'].write(json.dumps(metric_data))
        environ['wsgi.input'].seek(0)

        resp_body = self.app(environ, self.start_response)

        # Check response, smoke test.
        self.assertEqual(resp_body, [])
        self.assertEqual(self.resp_args[0].upper(), '200 OK')
        self.assertEqual(self.resp_args[1], [])

        # Check the metric was accepted.
        self.assertEqual(self.stats, [u'i.haz.clikd.cheezburgr:1|c'])

    def test_log_acceptance(self):
        message = 'Smoo clikd on a cheezburgr'
        metric_data = {
            'message': message,
            }
        environ = {
            'PATH_INFO': '/log',
            'REQUEST_METHOD': 'POST',
            }
        setup_testing_defaults(environ)
        # Assign the posted message.
        environ['wsgi.input'].write(json.dumps(metric_data))
        environ['wsgi.input'].seek(0)

        resp_body = self.app(environ, self.start_response)

        # Check response, smoke test.
        self.assertEqual(resp_body, [])
        self.assertEqual(self.resp_args[0], '200 OK')
        self.assertEqual(self.resp_args[1], [])

        # Check the metric was accepted.
        self.assertEqual([x.msg for x in self.logged], [message])

    def test_only_accepts_post(self):
        environ = {'REQUEST_METHOD': 'GET'}
        setup_testing_defaults(environ)

        resp_body = self.app(environ, self.start_response)

        # Check response, smoke test.
        self.assertEqual(resp_body, [])
        self.assertEqual(self.resp_args[0].upper(), '404 NOT FOUND')
        self.assertEqual(self.resp_args[1], [])

    def test_invalid_metric_type(self):
        metric_data = {
            'type': 'smudge',
            'label': 'doo.be.doo.be.do',
            'value': None,
            }
        environ = {
            'PATH_INFO': '/metric',
            'REQUEST_METHOD': 'POST',
            }
        setup_testing_defaults(environ)
        # Assign the posted message.
        environ['wsgi.input'].write(json.dumps(metric_data))
        environ['wsgi.input'].seek(0)

        from cnxlogging import InvalidMetricType
        with self.assertRaises(InvalidMetricType):
            resp_body = self.app.handle_metric(metric_data)

        resp_body = self.app(environ, self.start_response)
        self.assertEqual(self.resp_args[0], '500 Internal Server Error')
        self.assertEqual(self.resp_args[1], [('Content-type', 'text/plain')])
        self.assertEqual(resp_body,
                         ['InvalidMetricType: Invalid metric type: smudge'])


class StatsApplicationTestCase(unittest.TestCase):

    def setUp(self):
        self.log_handler = LogCapturingHandler(self)

        from cnxlogging import _StatsLoggingClient
        self.statist = _StatsLoggingClient()
        self.statist._logger.setLevel(logging.INFO)
        self.statist._logger.addHandler(self.log_handler)

        logger = logging.getLogger(self.__class__.__name__)
        logger.setLevel(logging.INFO)
        logger.addHandler(self.log_handler)

        from cnxlogging import Application
        self.app = Application(self.statist, logger)

    def start_response(self, *args, **kwargs):
        self.resp_args = args
        self.resp_kwargs = kwargs

    def test_stats_w_logging_client(self):
        # In the case where a statsd server has not been configured,
        #   stats info is sent to a log using a custom class that provides
        #   the same interface as the statsd.StatsClient.
        metric_data = {
            'type': 'timing',
            'label': 'i.haz.thunkd.cheezburgr',
            'value': 300,
            }
        environ = {
            'PATH_INFO': '/metric',
            'REQUEST_METHOD': 'POST',
            }
        setup_testing_defaults(environ)
        # Assign the posted message.
        environ['wsgi.input'].write(json.dumps(metric_data))
        environ['wsgi.input'].seek(0)

        resp_body = self.app(environ, self.start_response)

        # Check response, smoke test.
        self.assertEqual(resp_body, [])
        self.assertEqual(self.resp_args[0].upper(), '200 OK')
        self.assertEqual(self.resp_args[1], [])

        # Check the metric was accepted.
        self.assertEqual([x.msg for x in self.logged],
                         ['i.haz.thunkd.cheezburgr:300|ms'])


cnxlogging.py
# -*- coding: utf-8 -*-
# ###
# Copyright (c) 2013, Rice University
# This software is subject to the provisions of the GNU Affero General
# Public License version 3 (AGPLv3).
# See LICENCE.txt for details.
# ###
"""This is a simple interface for logging client-side metrics and messages."""
import logging
import json
import statsd


__all__ = ('Application', 'make_statist', 'paste_app_factory',)


STATS_LOGGER_NAME = 'stats'
SETTINGS_KEY__STATSD_HOST = 'statsd.host'
SETTINGS_KEY__STATSD_PORT = 'statsd.port'
SETTINGS_KEY__STATSD_PREFIX = 'statsd.prefix'

ACCEPTED_METRIC_TYPES = ('incr', 'gauge', 'timing',)

logger = logging.getLogger(__name__)


class BaseHandlingError(Exception):
    """Base exception for exceptions that happen inside
    the log and metric handlers.
    """


class InvalidMetricType(BaseHandlingError):

    def __init__(self, type_):
        message = 'Invalid metric type: {}'.format(type_)
        super(InvalidMetricType, self).__init__(message)


class _StatsLoggingClient(statsd.StatsClient):
    """This provides the same interface as statsd.StatsClient to make
    a logging compatible version of the statistics capturing methods.
    """

    def __init__(self, host=None, port=None, prefix=None, maxudpsize=512):
        """Keep the same parameters for easy instantiation."""
        self._addr = (host, port,)
        self._logger = logging.getLogger(STATS_LOGGER_NAME)
        self._prefix = prefix
        self._maxudpsize = maxudpsize

    def _send(self, data):
        """Send data to statsd."""
        self._logger.info(data.encode('ascii'))


def make_statist(settings):
    """Factory to create a statist object that will use statsd when
    configured or default to logging.
    """
    host = settings.get(SETTINGS_KEY__STATSD_HOST, None)
    port = settings.get(SETTINGS_KEY__STATSD_PORT, 8125)
    prefix = settings.get(SETTINGS_KEY__STATSD_PREFIX, None)

    # Is statsd configured?
    if host is not None:
        # Use statsd client.
        klass = statsd.StatsClient
    else:
        # Use statsd logging clone.
        klass = _StatsLoggingClient
    statist = klass(host, port, prefix)
    return statist


class Application:
    """Metric and message logging application.
    The given ``statist`` should be a ``statsd.StatsClient``
    or object with a compatible interface.
    The ``logger`` is a stand python logger from ``logging.getLogger``.
    The ``path`` is an optional path to listen on.
    If ``path`` is ``None``, the application will handle
    requests coming in to any path.
    """

    def __init__(self, statist, logger, path=None):
        self.statist = statist
        self.logger = logger

    def __call__(self, environ, start_response):
        if environ['REQUEST_METHOD'] != 'POST':
            return self.not_found(environ, start_response)

        # Routing (poorly, but simply done).
        if environ['PATH_INFO'] == '/metric':
            handler = self.handle_metric
        elif environ['PATH_INFO'] == '/log':
            handler = self.handle_log
        else:
            return self.not_found(environ, start_response)

        # Handle errors and don't send the entire traceback.
        try:
            handler(self._parse_message_body(environ))
        except Exception as exc:
            start_response('500 Internal Server Error',
                           [('Content-type', 'text/plain')])
            resp = ["{}: {}".format(exc.__class__.__name__, exc.message)]
        else:
            start_response('200 OK', [])
            resp = []

        return resp

    def not_found(self, environ, start_response):
        start_response('404 Not Found', [])
        return []

    def _parse_message_body(self, environ):
        return json.load(environ['wsgi.input'])

    def handle_log(self, payload):
        self.logger.info(payload['message'])

    def handle_metric(self, payload):
        metric_type = payload['type']
        if metric_type not in ACCEPTED_METRIC_TYPES:
            raise InvalidMetricType(metric_type)
        method = getattr(self.statist, metric_type)

        label = payload['label']
        value = payload.get('value', None)
        value = value is None and 1 or value
        method(label, value)
        

def paste_app_factory(global_config, **local_config):
    """Makes a WSGI application using the ``PasteDeploy`` interface."""
    statist = make_statist(local_config)
    logger = logging.getLogger(__name__)
    return Application(statist, logger)
